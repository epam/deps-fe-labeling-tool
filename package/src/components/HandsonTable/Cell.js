
import { COLORS } from '@/theme/theme.default'

const renderCell = (td, value, cellProps, extra) => {
  cellProps.readOnly && (td.style.color = COLORS.GRAYSCALE_1)
  td.textContent = value
  extra && td.append(extra)
  return td
}

export {
  renderCell
}
