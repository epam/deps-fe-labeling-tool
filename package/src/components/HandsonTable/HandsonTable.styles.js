/* stylelint-disable */
import 'handsontable/dist/handsontable.full.css'
import { HotTable } from '@handsontable/react'
import styled from 'styled-components'

export const StyledHotTable = styled(HotTable)`
  position: relative;

  .ht_master {
    overflow: visible !important;
  }

  .handsontableInputHolder,
  && .wtBorder.current,
  && .wtBorder.area {
    z-index: unset;
  }

  .ht_clone_top,
  .ht_clone_left,
  .ht_clone_top_left_corner {
    z-index: 1;
  }

  .handsontableInput {
    box-sizing: content-box;
  }
`
