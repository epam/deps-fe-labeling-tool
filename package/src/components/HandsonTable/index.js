
export * from './models'
export * from './HandsonTable'
