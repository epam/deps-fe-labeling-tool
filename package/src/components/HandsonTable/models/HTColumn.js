
import PropTypes from 'prop-types'

class HTColumn {
  constructor (data) {
    this.data = data
  }
}

const htColumnShape = PropTypes.shape({
  data: PropTypes.string.isRequired
})

export {
  HTColumn,
  htColumnShape
}
