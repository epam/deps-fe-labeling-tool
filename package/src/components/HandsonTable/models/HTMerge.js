import PropTypes from 'prop-types'

class HTMerge {
  constructor (row, col, rowspan, colspan) {
    this.row = row
    this.col = col
    this.rowspan = rowspan
    this.colspan = colspan
  }
}

const htMergeShape = PropTypes.shape({
  row: PropTypes.number.isRequired,
  col: PropTypes.number.isRequired,
  rowspan: PropTypes.number.isRequired,
  colspan: PropTypes.number.isRequired
})

export {
  HTMerge,
  htMergeShape
}
