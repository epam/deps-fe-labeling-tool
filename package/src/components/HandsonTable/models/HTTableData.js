
import PropTypes from 'prop-types'

const rowOfPrimitivesShape = PropTypes.arrayOf(
  PropTypes.oneOfType([
    PropTypes.number.isRequired,
    PropTypes.string.isRequired
  ])
)

const rowsOfPrimitivesShape = PropTypes.arrayOf(
  rowOfPrimitivesShape
)

export { rowsOfPrimitivesShape }
