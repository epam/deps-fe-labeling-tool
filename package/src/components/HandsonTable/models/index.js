
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
