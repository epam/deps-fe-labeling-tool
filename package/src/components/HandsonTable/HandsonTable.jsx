
import React, { createRef, Component, memo } from 'react'
import PropTypes from 'prop-types'
import { renderCell } from './Cell'
import { StyledHotTable } from './HandsonTable.styles'
import { htColumnShape } from './models/HTColumn'
import { htMergeShape } from './models/HTMerge'
import { rowsOfPrimitivesShape } from './models/HTTableData'

const StretchColumn = {
  ALL: 'all',
  LAST: 'last',
  NONE: 'none'
}

const EventSource = {
  EDIT: 'edit',
  LOAD_DATA: 'loadData',
  POPULATE_FROM_ARRAY: 'populateFromArray'
}

const HORIZONTAL_SCROLL_HEIGHT = 10
const HEIGHT_DIFFERENCE_THRESHOLD = 1

class HandsonTable extends Component {
  static defaultProps = {
    colHeaders: true,
    manualColumnResize: true,
    manualRowResize: true,
    outsideClickDeselects: true,
    rowHeaders: true,
    stretchH: StretchColumn.ALL
  }

  static propTypes = {
    cellRenderer: PropTypes.func,
    colHeaders: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.arrayOf(PropTypes.string)
    ]),
    readOnly: PropTypes.bool,
    renderExtra: PropTypes.func,
    rowHeaders: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.arrayOf(PropTypes.string)
    ]),
    manualColumnResize: PropTypes.bool,
    manualRowResize: PropTypes.bool,
    outsideClickDeselects: PropTypes.bool,
    stretchH: PropTypes.oneOf(
      Object.values(StretchColumn)
    ),
    saveData: PropTypes.func,
    data: rowsOfPrimitivesShape.isRequired,
    className: PropTypes.string,
    mergeCells: PropTypes.arrayOf(
      htMergeShape
    ).isRequired,
    columns: PropTypes.arrayOf(
      htColumnShape
    )
  }

  htRef = createRef()

  ht = null

  autoRowSizePlugin = null

  getWithoutPx = (computedValue) => Number(computedValue.slice(0, -2))

  setHeight = () => {
    const hotElementRef = this.htRef.current?.hotElementRef

    const partentComputedStyle = window.getComputedStyle(hotElementRef.parentNode)
    const parentPaddingTop = this.getWithoutPx(partentComputedStyle.paddingTop)
    const parentPaddingBottom = this.getWithoutPx(partentComputedStyle.paddingBottom)

    this.autoRowSizePlugin?.recalculateAllRowsHeight()

    const headerHeight = this.autoRowSizePlugin?.getColumnHeaderHeight() || 0
    const tableHeight = this.autoRowSizePlugin?.heights
      .reduce((acc, current) => acc + current, headerHeight)
    const wrapperHeight = hotElementRef?.offsetHeight
    const parentHeight = hotElementRef?.parentNode.offsetHeight - parentPaddingTop - parentPaddingBottom

    if (!tableHeight || !wrapperHeight) {
      return
    }

    const targetHeight = tableHeight + HORIZONTAL_SCROLL_HEIGHT

    if (Math.abs(wrapperHeight - targetHeight) <= HEIGHT_DIFFERENCE_THRESHOLD) {
      return
    }

    this.ht.updateSettings({
      height: Math.min(Math.abs(parentHeight), targetHeight)
    })
  }

  onAfterChange = (changes, source) => {
    if (!changes) {
      return
    }

    const [necessaryChanges] = changes

    if (source === EventSource.EDIT) {
      this.props.saveData(necessaryChanges)
    }
  }

  cellRenderer = (instance, td, row, col, prop, value, cellProps) => {
    if (this.props.cellRenderer) {
      return this.props.cellRenderer(instance, td, row, col, prop, value, cellProps)
    }
    const extra = this.props.renderExtra && this.props.renderExtra(row, col, this.ht?.getSourceData())
    return renderCell(td, value, cellProps, extra)
  }

  componentDidMount = () => {
    this.ht = this.htRef.current.hotInstance
    this.autoRowSizePlugin = this.ht.getPlugin('AutoRowSize')

    this.setHeight()
  }

  componentWillUnmount = () => {
    this.ht.deselectCell()
  }

  render = () => (
    <StyledHotTable
      ref={this.htRef}
      afterChange={this.onAfterChange}
      className={this.props.className}
      colHeaders={this.props.colHeaders}
      columns={this.props.columns}
      data={this.props.data}
      manualColumnResize={this.props.manualColumnResize}
      manualRowResize={this.props.manualRowResize}
      mergeCells={this.props.mergeCells}
      outsideClickDeselects={this.props.outsideClickDeselects}
      readOnly={this.props.readOnly}
      renderer={this.cellRenderer}
      rowHeaders={this.props.rowHeaders}
      stretchH={this.props.stretchH}
    />
  )
}

const MemoHandsonTable = memo(HandsonTable)

export { MemoHandsonTable as HandsonTable }
