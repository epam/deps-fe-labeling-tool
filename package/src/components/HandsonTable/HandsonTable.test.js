/* eslint-disable jest/no-conditional-expect */

import { mockUuid } from '@/mocks/mockUuid'
import React from 'react'
import { shallow } from 'enzyme'
import { Table, CellsMerge, CellValue } from '@/models/Table'
import { StyledHotTable } from './HandsonTable.styles'
import {
  HandsonTable,
  HTMerge,
  HTColumn
} from '.'

jest.mock('uuid', () => mockUuid)

const mockData = new Table(
  [10, 20, 30, 40],
  [50, 60, 70, 80],
  [new CellsMerge(1, 0, 1, 2)],
  [new CellValue(0, 0, '', 0.3), new CellValue(1, 0, 'Piloncillo D', 0.8)],
  {},
  'codeTable',
  1,
  1
)

const mockChangedCellData = [
  [
    1,
    2,
    'prevValue',
    'newValue'
  ]
]

const mockSourceData = [
  ['test'],
  ['test1']
]
const mockMergedCells = []
const mockExtra = <div>mock extra box</div>

Object.defineProperty(window, 'getComputedStyle', {
  value: () => ({
    paddingTop: '5px',
    paddingBottom: '5px'
  })
})

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  createRef: () => ({
    current: {
      hotInstance: {
        getSettings: jest.fn(),
        updateSettings: jest.fn(),
        getSourceData: jest.fn(() => mockSourceData),
        selectCells: jest.fn(),
        renderExtra: jest.fn(() => mockExtra),
        scrollViewportTo: jest.fn(),
        getPlugin: jest.fn(() => ({
          mergedCellsCollection: {
            mergedCells: mockMergedCells
          },
          enablePlugin: jest.fn(),
          disablePlugin: jest.fn(),
          recalculateAllRowsHeight: jest.fn(),
          getColumnHeaderHeight: jest.fn(() => 25),
          heights: [23, 23, 23]
        }))
      },
      hotElementRef: {
        offsetHeight: 114,
        parentNode: {
          paddingTop: '5px',
          paddingBottom: '5px'
        }
      }
    }
  })
}))

jest.mock('./Cell', () => ({
  renderCell: jest.fn()
}))

describe('Component: HandsonTable', () => {
  let defaultProps
  let wrapper
  let htProps
  const EventSource = {
    EDIT: 'edit',
    LOAD_DATA: 'loadData',
    POPULATE_FROM_ARRAY: 'populateFromArray'
  }

  beforeEach(() => {
    jest.clearAllMocks()

    defaultProps = {
      activePage: 1,
      onChangeData: jest.fn(),
      columns: [
        new HTColumn('0.value'),
        new HTColumn('1.value')
      ],
      data: mockData,
      mergeCells: [
        new HTMerge(0, 1, 2, 1)
      ],
      readOnly: false,
      saveData: jest.fn(),
      renderExtra: jest.fn()
    }

    wrapper = shallow(<HandsonTable {...defaultProps} />)
    htProps = wrapper.find(StyledHotTable).props()
  })

  it('should render correct layout', () => {
    expect(wrapper.dive()).toMatchSnapshot()
  })

  it('should set readonly to true in case of props.readOnly is true', () => {
    defaultProps.readOnly = true
    wrapper.setProps(defaultProps)
    const { readOnly } = wrapper.find(StyledHotTable).props()
    expect(readOnly).toEqual(true)
  })

  it('should call props.saveData only in case of EventSource is `edit` and calling afterChange handsonTable hook', () => {
    Object.values(EventSource).forEach((s) => {
      htProps.afterChange(mockChangedCellData, s)

      if (s === EventSource.EDIT) {
        expect(defaultProps.saveData).nthCalledWith(1, mockChangedCellData[0])
        jest.clearAllMocks()
      } else {
        expect(defaultProps.saveData).not.toHaveBeenCalled()
      }
    })
  })
})
