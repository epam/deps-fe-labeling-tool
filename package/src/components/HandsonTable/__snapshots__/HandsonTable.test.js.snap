// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: HandsonTable should render correct layout 1`] = `
<HotTable
  afterChange={[Function]}
  className="HandsonTablestyles__StyledHotTable-sc-14xqz4u-0 bPkPFr"
  columns={
    Array [
      HTColumn {
        "data": "0.value",
      },
      HTColumn {
        "data": "1.value",
      },
    ]
  }
  data={
    Table {
      "fieldCode": "codeTable",
      "index": 1,
      "merges": Array [
        CellsMerge {
          "colspan": 1,
          "column": 0,
          "row": 1,
          "rowspan": 2,
        },
      ],
      "meta": Object {},
      "typeName": "table",
      "uid": 1,
      "values": Array [
        CellValue {
          "column": 0,
          "confidence": 0.3,
          "row": 0,
          "value": "",
        },
        CellValue {
          "column": 0,
          "confidence": 0.8,
          "row": 1,
          "value": "Piloncillo D",
        },
      ],
      "xGuidelines": Array [
        10,
        20,
        30,
        40,
      ],
      "yGuidelines": Array [
        50,
        60,
        70,
        80,
      ],
    }
  }
  mergeCells={
    Array [
      HTMerge {
        "col": 1,
        "colspan": 1,
        "row": 0,
        "rowspan": 2,
      },
    ]
  }
  readOnly={false}
  renderer={[Function]}
/>
`;
